### DISCUSSION TITLE
How to get count of child records in hasMany relationship

### DISCUSSION THREAD

**Discussion by HarshalNathe** [loopbackio/loopback-next#7345] [LB3 Migration] [answered]
Currently I'm using the loopback3, and I'm planing to migrate my project into Loopback4. In my existing project I have created the relational API's using hasMany relationship. In loopback3 I'm able to get the count of child table once I pass the ID of parent table record(and also able to filter the count by where condition).

**Scenario:** Tenant has many users. when I created hasMany relationship in loopback3, it provides me the `GET API` to get the count of users by tenant ID. (Counts users of tenant)

**E.x.** `GET /tenants/{id}/users/count`

But when I created hasMany relationship in Loopback4 I'm not getting API like this in my OpenAPI swagger(in tenant-user controller). Is there any workaround for this? OR is there any way to achieve this scenario in Loopback4? 

I'm using **MySQL** as my **datasource**.

**Comment by marioestradarosa** [comment, marked as answer]
@HarshalNathe ,  

Indeed that /count path is not created by default. But you can add it easily as follows:  I will take the `todo-list` example as the basis for my explanation. The key part here is that the `count` method resides in the child repository. The controller that contains the relationship is `todo-list-todo.controller.ts` so we need to import the repository of the child first, since the constrained relation repository **_does not have this method available for now_**.

 Also the count method from the repository receives a Where parameter, so we need to force the parent ID, so it behaves similar to other `/count` in regular controllers.

**Step 1:**
   Import the repository of the child in the relationship and the `WhereBuilder`  from the @loopback/repository package.
```ts
import { Count, CountSchema, Filter, repository,Where,WhereBuilder } from '@loopback/repository';
import { TodoListRepository, TodoRepository } from '../repositories';
```
**Step 2:**
**Inject** the `todo` repository in the controller's constructor .
```ts
 @repository(TodoRepository)
 protected todoRepository: TodoRepository
```

**Step 3:**
Add the new controller method according to your requirement as follows:
```ts
@get('/todo-lists/{id}/todos/count', {
    responses: {
      '200': {
        description: 'Todo-list todo model count',
        content: { 'application/json': { schema: CountSchema } },
      },
    },
  })
  async todoListTodoCount(
    @param.path.number('id') id: number,
    @param.where(Todo) where?: Where<Todo>,
  ): Promise<Count> {
    const imposedWhereByTodoList = { todoListId: id };

    if (where) {
      where = new WhereBuilder(where).and(imposedWhereByTodoList).build();
    } else {
      where = new WhereBuilder(imposedWhereByTodoList).build();

    }
    return this.todoRepository.count(where);
  }
```


**Additionally/Optionally** and not related to the question in this discussion, if someone is trying to run at this moment the `todo-list` from `lb4 examples` , there will be an error in the unit testing, to correct it, let's do the following correction in the file `todo-list-todo.controller.unit.ts` file.

_Add the todo repository:_
``` import { TodoListRepository, TodoRepository } from '../../../repositories'; ```

In the first `describe` below the `let todoListRepo:`  add the following: 
`  let todoRepo: StubbedInstanceWithSinonAccessor<TodoRepository>;` 

Now in the last line change this line from 
`    controller = new TodoListTodoController(todoListRepo); `
To
`    controller = new TodoListTodoController(todoListRepo, todoRepo); `

And you will have it running without any error.

**Reply by HarshalNathe** [reply]
Thanks, @marioestradarosa this seems like the proper way to do things. And this helps me to achieve the mentioned scenario. 
Before your response, I also tried the below code, which is also helping me to achieve the same result.

```javascript
  @get('/tenants/{id}/users/count', {
    responses: {
      '200': {
        description: 'Count of Tenant has many User',
        content: {
          'application/json': {
            schema: {type: 'array', items: getModelSchemaRef(User)},
          },
        },
      },
    },
  })
  async count(
    @param.path.number('id') id: number,
    @param.where(User) where?: Where<User>,
  ): Promise<Count> {
    return this.userRepository.count({tenantId: id, ...where});
  }
```

**Reply by marioestradarosa** [reply]
@HarshalNathe nice work!, keep in mind that you have wrong the OpenAPI spec for the response. This might not affect the code execution but whenever you generate an openapi spec json file and share with a tester it will not reflect the count response.  so the proper response is : `    content: { 'application/json': { schema: CountSchema } }, `

**Reply by HarshalNathe** [reply]
Yes @marioestradarosa got it. I have quick question for you. Now the same case is happening with below API as well. 

```
GET 	tenant/{id}/users/{fk} | Find a related item by id for users
DELETE  tenant/{id}/users/{fk} | Delete a related item by id for users
PUT 	tenant/{id}/users/{fk} | Update a related item by id for users

```
So My question is that, Can I follow the same way to implement these API's too?

**Reply by marioestradarosa** [reply]
@HarshalNathe , in the `lb4 example  todo-list`. the `GET`, `DELETE` and `PATCH` are already generated for you. Do these generated controller does not satisfy your requirements ?. Notice that in this case it is referencing the **_HasManyRepository_** instead of going directly with the `**_TodoRepository_**` (which we had to use, because count method is not already implemented in the HasManyRepository yet, but a PR is coming for that.)

Example for the `GET`.  You can send a filter to see if that satisfy you, otherwise, feel free to implement it in the similar way as the `count` . Remember that is exactly the beauty in `Loopback 4` , it is very flexible, not opinionated.

```ts
@get('/todo-lists/{id}/todos', {
    responses: {
      '200': {
        description: 'Array of TodoList has many Todo',
        content: {
          'application/json': {
            schema: { type: 'array', items: getModelSchemaRef(Todo) },
          },
        },
      },
    },
  })
  async find(
    @param.path.number('id') id: number,
    @param.query.object('filter') filter?: Filter<Todo>,
  ): Promise<Todo[]> {
    return this.todoListRepository.todos(id).find(filter);
  }

```
